### PR TITLE
Decimal functions

### PR DESCRIPTION
**Pull request by fiedukow**
This is set of commits introducing decimal version of various functions

### CODE REVIEW

**Comment by martint** [comment]
@haozhun, can you review?

**Comment by haozhun** [comment]
@martint, I'll review. Expect a lot of functions from me.

**Comment by fiedukow** [comment]
ping

**Comment by fiedukow** [comment]
https://groups.google.com/forum/#!topic/presto-users/H9J9dgiY2xk
ping

**Comment by fiedukow** [comment]
Rebased to master, conflicts solved.

**Comment by KBP-TDC** [comment]
@haozhun  @martint : this PR blocks aggregates for DECIMAL (see presto-users group posts).

**Comment by fmeiser** [comment]
A `ShortDecimal ` is limited to precision of 17 digits but `long ` could contain 18 digits.
What is the reason? Other engines use short upto 18 digits.

**Comment by ilfrin** [comment]
Please ask this question on the user group not in this PR :). You have better chances getting an answer there as we'd need to dig out the discussions back when we were implementing it.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestLongDecimalAverageAggregation.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import java.math.BigDecimal;
+import java.util.List;
+
+import static com.facebook.presto.spi.type.Decimals.MAX_SHORT_PRECISION;
+import static com.facebook.presto.spi.type.Decimals.encodeScaledValue;
+
+public class TestLongDecimalAverageAggregation
+        extends AbstractTestDecimalAverageAggregation
+{
+    DecimalType longDecimalType = DecimalType.createDecimalType(MAX_SHORT_PRECISION + 1);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestLongDecimalAverageAggregation.java]: `private final`

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestShortDecimalAverageAggregation.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import static com.facebook.presto.spi.type.Decimals.MAX_SHORT_PRECISION;
+
+public class TestShortDecimalAverageAggregation
+        extends AbstractTestDecimalAverageAggregation
+{
+    DecimalType shortDecimalType = DecimalType.createDecimalType(MAX_SHORT_PRECISION - 1);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestShortDecimalAverageAggregation.java]: `private static final`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.math.BigDecimal.ROUND_HALF_UP;
+
+public class DecimalAverageAggregation
+        extends SqlAggregationFunction
+{
+    public static final DecimalAverageAggregation DECIMAL_AVERAGE_AGGREGATION = new DecimalAverageAggregation();
+    private static final String NAME = "avg";
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: Empty line before this line

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: I got myself really confused when I was trying to understand the correctness of this method. Your `deserialize` implementation suggests that `otherState.getBigInteger()` can be null. However, such case is not handled here and yet this code passes all tests.

I dug into it. It turns out there were some unintended guarantees provided by the implementation that calls `combine`. I removed those guarantees in #5637. I expect #5637 to be merged by the end of day today so that you can rebase this pull request on top of master. There will be test failures in your commits as a result. Please fix those failures.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/BigIntegerAndLongStateFactory.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        @Override
+        public void setLong(long value)
+        {
+            longs.set(getGroupId(), value);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return bigIntegers.sizeOf() + longs.sizeOf();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/BigIntegerAndLongStateFactory.java]: This is the shallow size, which is wrong. The actual memory usage can be multiple times more than estimated. This can result in crash of the cluster when queries like `SELECT sum(y) FROM t GROUP BY x`, where `x` is unique and `y` is decimal.

Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/BigIntegerAndLongStateFactory.java]: I looked at `BigInteger` implementation, the combination of shallow size of `BigInteger` and 4 * `BigInteger.mag.length` is what you want. 

To get the accurate shallow size of a class, you can use `ClassLayout.parseClass(BigInteger.class).instanceSize()` which is provided by http://openjdk.java.net/projects/code-tools/jol/.

To get `BigInteger.mag.length`, there are two ways. You can either:
* subclass `BigInteger`
* use `MethodHandle.unreflectGetter` and reflection

Alternatively, a good enough estimate here might be 
* shallow size + ceil(precision of the column type, 4) + 4

Out of the 3 possibilities I proposed above, I'll need to talk to @martint and see what he thinks.

**Comment by haozhun** [comment]
I had some comments for the 1st commit (average). Some of those comments probably also apply to other commits. Feel free to ping me after addressing the comments.

**Comment by haozhun** [comment]
Please hold off addressing the comment just yet. I'll let you know.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java (hunk 4 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: @haozhun why couldn't we change this after merging in our changes? As you see other classes behave similarly to this one (you change them in your PR), as I see this PR does not breaking anything, it's working within what is established today. Then after we land this we could safely change the assumptions about the State objects that you don't like in the framework. We've been rebasing this PR for some time now as time goes by, so I think if this is not a show-stopper we should follow the chronological order here. Thanks.

**Comment by haozhun** [comment]
@ilfrin @fiedukow Sorry for the confusion I have caused. I deleted my comments that @ilfrin raised opinion against. @ilfrin has a valid point.

Please go ahead and address any comments that are still there.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestLongDecimalAverageAggregation.java (hunk 1 above)
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestLongDecimalAverageAggregation.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestShortDecimalAverageAggregation.java (hunk 2 above)
Comment (ilfrin) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestShortDecimalAverageAggregation.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java (hunk 3 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: done

**Comment by ilfrin** [comment]
@haozhun I was thinking about it ... how about I add/implement the `addMemoryUsage(long)` method to the state interface/class, following the pattern I see in other `State` classes that consist of non-primitives. Then I would call it in `combine()` and then use it to output the estimated size without those implementation specific nuances?

isn't that simpler and better?

**Comment by ilfrin** [comment]
@haozhun, we can do an even simpler and clearer approach (following what other `AccumulatorStateFactory` implementation do):

lets have `GroupedBigIntegerAndLongState.getEstimatedSize()` just do 
```
return bigIntegers.sizeOf() + longs.sizeOf() + estimatedSizeOfBigIntegerObjects + SIZE_OF_LONG;
```
where `estimatedSizeOfBigIntegerObjects` is collected like this

```
       GroupedBigIntegerAndLongState.setBigInteger(BigInteger value)
        {
            requireNonNull(value, "value is null");
            if (getBigInteger() == null) {
                estimatedSizeOfBigIntegerObjects += BIG_INTEGER_APPROX_SIZE;
            }
            bigIntegers.set(getGroupId(), value);
        }
```
where `BIG_INTEGER_APPROX_SIZE = 4 * 16 + 64`, which is the upper limit for our case.

This should account for everything and is very clear and simple.

**Comment by ilfrin** [comment]
I think the LongBigArray class (`longs.sizeOf()`) properly accounts for it's size, so this part is fine.

**Comment by haozhun** [comment]
@ilfrin Using JOL is really easy. It's just one line. You can take a look at the `INSTANCE_SIZE` field of `VariableWidthBlock` for example.

The overhead of `BigInteger` is the sum of the aforementioned `INSTANCE_SIZE` and `Sizeof.sizeOf(new int[0])`.

Now, whether 16 byte is a good enough estimation of the content of `mag` field of `BigInteger`. I believe you are right. It's probably good  enough. `ClassLayout.parseClass(BigInteger.class).instanceSize() + SizeOf.sizeOf(new int[0])` produces 56 on my laptop with compressed pointer. Whether the `mag` should be 8, 12, 16, or 20 probably doesn't matter much.

Use 
`ClassLayout.parseClass(BigInteger.class).instanceSize() + SizeOf.sizeOf(new int[0]) + 5 * SIZE_OF_INT` for your `BIG_INTEGER_APPROX_SIZE`. (I'm using 5 instead of 4 just to be conservative. The sum part of an average aggregation can be more than 128-bit. It would be nice to have comment on this.)

Also, make sure you not only check the case where null becomes non-null, but also the other way around.

Otherwise, your solution for memory accounting is good.

**Comment by ilfrin** [comment]
@haozhun, sorry for not attending this, that's fine, thanks I'll use that one-liner instead then and add the comment. 

I'm not sure though I understand the case you mention last "but also the other way around", where is the BigInteger for a group being nulled? The setBigInteger has a requireNonNull at the very top to prevent that. Or am I missing something?

**Comment by ilfrin** [comment]
@haozhun applied the comments so far (mod the question I had)

**Comment by ilfrin** [comment]
@haozhun please review this further, this is has been waiting for long and now we're working on the fast decimal implementation which is based on this, I'd rather this lands ASAP to ease the pain of having multiple layers of interdependent pull requests and rebasing them all the time. Thanks! :) 

cc @martint

**Comment by losipiuk** [comment]
@martint, @haozhun I rebased it to current master.
Please take a look at that when you have a chance. 
We already have much faster decimal implementation waiting but we wanted to land this first to the latter one 
only changes internal implementation details leaving overall concept intact.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    private static final MethodHandle LONG_DECIMAL_INPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "inputLongDecimal", Type.class, BigIntegerAndLongState.class, Block.class, int.class);
+
+    private static final MethodHandle SHORT_DECIMAL_OUTPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "outputShortDecimal", DecimalType.class, BigIntegerAndLongState.class, BlockBuilder.class);
+    private static final MethodHandle LONG_DECIMAL_OUTPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "outputLongDecimal", DecimalType.class, BigIntegerAndLongState.class, BlockBuilder.class);
+
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(DecimalAverageAggregation.class, "combine", BigIntegerAndLongState.class, BigIntegerAndLongState.class);
+
+    public DecimalAverageAggregation()
+    {
+        super(NAME, ImmutableList.of(Signature.withVariadicBound("T", DECIMAL)), ImmutableList.of(), parseTypeSignature("T"), ImmutableList.of(parseTypeSignature("T")));
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: @martint Do we generally care whether we use
* T of variadic bound decimial, or
* decimal(p,s)

It feels like the latter is better because the variadic bound was created primarily for types like `row

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    private static final MethodHandle LONG_DECIMAL_OUTPUT_FUNCTION = methodHandle(DecimalSumAggregation.class, "outputLongDecimal", DecimalType.class, BigIntegerState.class, BlockBuilder.class);
+
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(DecimalSumAggregation.class, "combine", BigIntegerState.class, BigIntegerState.class);
+
+    public DecimalSumAggregation()
+    {
+        super(NAME,
+                ImmutableList.of(),
+                ImmutableList.of(),
+                parseTypeSignature("decimal(38,s)", ImmutableSet.of("s")),
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalSumAggregation.java]: @martint This appears like the only reasonable return type given the current system we have.  But I want to see what you think about it.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java (hunk 6 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java]: Yes, we should use decima(p,s). Variadic bounds is legacy, and we're probably going to remove it at some point.

**Comment by losipiuk** [comment]
@haozhun I dropped usage of variadicBound with d637318

**Comment by ilfrin** [comment]
rebased and squashed fixups

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java [hunk 8]
```diff
@@ -65,19 +71,28 @@ public void setup()
     @Test
     public void testAllRegistered()
     {
-        Set<Type> orderableTypes = METADATA.getTypeManager()
-                .getTypes().stream()
+        Set<Type> orderableTypes = getTypes().stream()
                 .filter(Type::isOrderable)
-                .collect(toImmutableSet());
+                .collect(toSet());
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java]: This line is unrelated change.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 9]
```diff
@@ -174,6 +196,60 @@ public static double ceiling(@SqlType(StandardTypes.DOUBLE) double num)
...
+                .longVariableConstraints(longVariableExpression("return_precision", "num_precision - num_scale + min(num_scale, 1)"))
+                .argumentTypes(parseTypeSignature("decimal(num_precision,num_scale)", ImmutableSet.of("num_precision", "num_scale")))
+                .returnType(parseTypeSignature("decimal(return_precision,0)", ImmutableSet.of("return_precision")))
+                .build();
+        return SqlScalarFunction.builder(MathFunctions.class)
+                .signature(signature)
+                .implementation(b -> b
+                        .methods("ceilingShortShortDecimal")
+                        .withExtraParameters(MathFunctions::decimalTenToScaleAsLongExtraParameters)
+                )
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: In Presto code, we generally put `)` at the end of the previous line.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 10]
```diff
@@ -174,6 +196,60 @@ public static double ceiling(@SqlType(StandardTypes.DOUBLE) double num)
...
+    {
+        return ImmutableList.of(bigIntegerTenToNth(context.getLiteral("num_scale").intValue()));
+    }
+
+    private static List<Object> decimalTenToScaleAsLongExtraParameters(SpecializeContext context)
+    {
+        return ImmutableList.of(longTenToNth(context.getLiteral("num_scale").intValue()));
+    }
+
+    public static long ceilingShortShortDecimal(long num, long divisor)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Add `@UsedByGeneratedCode`

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java [hunk 11]
```diff
@@ -247,7 +246,7 @@ public void testMinLongDecimalDecimal()
         assertAggregation(
                 function,
                 1.0,
-                sqlDecimal("2.2"),
+                SqlDecimal.of("2.2"),
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java]: Put changes to TestMinMaxByAggregation in `"min_by", "max_by" aggregate functions for decimal(p,s)`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 12]
```diff
@@ -152,12 +154,28 @@ public void testCeil()
         assertFunction("ceiling(-123.0)", DOUBLE, -123.0);
         assertFunction("ceiling(123.45)", DOUBLE, 124.0);
         assertFunction("ceiling(-123.45)", DOUBLE, -123.0);
-        assertFunction("ceiling(CAST(NULL AS DOUBLE))", DOUBLE, null);
         assertFunction("ceiling(REAL '123.0')", REAL, 123.0f);
         assertFunction("ceiling(REAL '-123.0')", REAL, -123.0f);
         assertFunction("ceiling(REAL '123.45')", REAL, 124.0f);
         assertFunction("ceiling(REAL '-123.45')", REAL, -123.0f);
+
+        assertFunction("ceiling(DECIMAL '123')", createDecimalType(3), SqlDecimal.of("123"));
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: Split these tests into 3 sections (separated by empty lines), each covering one of short-short, long-short, long-long.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 13]
```diff
@@ -245,12 +263,29 @@ public void testFloor()
...
         assertFunction("floor(123.45)", DOUBLE, 123.0);
         assertFunction("floor(-123.45)", DOUBLE, -124.0);
+
         assertFunction("floor(REAL '123.0')", REAL, 123.0f);
         assertFunction("floor(REAL '-123.0')", REAL, -123.0f);
         assertFunction("floor(REAL '123.45')", REAL, 123.0f);
         assertFunction("floor(REAL '-123.45')", REAL, -124.0f);
-        assertFunction("floor(CAST(NULL as DOUBLE))", DOUBLE, null);
+
+        assertFunction("floor(DECIMAL '123')", createDecimalType(3), SqlDecimal.of("123"));
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: Split these tests into 3 sections (separated by empty lines), each covering one of short-short, long-short, long-long.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 14]
```diff
@@ -532,6 +652,217 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
         return Math.round(num * factor) / factor;
     }
 
+    private static SqlScalarFunction decimalRoundFunction()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: round, truncate, mod, and sign should not be in this commit.

**Comment by haozhun** [comment]
Can we remove the `Conflicts:` sections in commit messages?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 10 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Do the same for other functions that are similar.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 15]
```diff
@@ -96,8 +97,8 @@ public static long absInteger(@SqlType(StandardTypes.INTEGER) long num)
 
     @Description("absolute value")
     @ScalarFunction
-    @SqlType(StandardTypes.BIGINT)
-    public static long abs(@SqlType(StandardTypes.BIGINT) long num)
+    @SqlType(BIGINT)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: Do not static import `StandardTypes.*`.

**Comment by haozhun** [comment]
* `avg`, `sum`, `checksum`, `abs` are good.
* `min_by`, `max_by`, `ceiling`, `floor`, `sign` are good except the commits are messed up. (Changes that should be there are in other commits, or vice versa.)
* I haven't looked at `mod`, `round`, `truncate` because the commits are too messed up.
* `Add tests for decimal aggregate functions state serializers` is good.

Please apply my comments and sort out the commits.

Additionally, please let me know whether you want changes that are good merged first.

**Comment by haozhun** [comment]
You only have `truncateN` but not `truncate`. Is this intentional?

You only have 2 versions instead of 3 for `round` (not `roundN`). This seems a bug.

File: presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java [hunk 16]
```diff
@@ -469,27 +470,37 @@ private static BigInteger internalDivideDecimals(BigInteger aBigInteger, BigInte
...
+    public static SignatureBuilder modulusSignatureBuilder()
+    {
         TypeSignature decimalLeftSignature = parseTypeSignature("decimal(a_precision, a_scale)", ImmutableSet.of("a_precision", "a_scale"));
         TypeSignature decimalRightSignature = parseTypeSignature("decimal(b_precision, b_scale)", ImmutableSet.of("b_precision", "b_scale"));
         TypeSignature decimalResultSignature = parseTypeSignature("decimal(r_precision, r_scale)", ImmutableSet.of("r_precision", "r_scale"));
 
-        Signature signature = Signature.builder()
-                .kind(SCALAR)
-                .operatorType(MODULUS)
+        return Signature.builder()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java]: It seems you can leave `.kind(SCALAR)` here.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 14 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: round, truncate and sign got mixed up here during some fixup squashing ... I will untangle this, but mod is already in it's own commit, so that should be OK

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java (hunk 8 above)
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestMinMaxByAggregation.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 9 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: changed

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 10 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 12 above)
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: done

Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 13 above)
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 15 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

**Comment by losipiuk** [comment]
> You only have truncateN but not truncate. Is this intentional?

I do not believe this was intentional. But we already have followup patch - which operates on 
newer/better/faster representation of decimal values which adds the missing one so let's skip this one
as it would be overwritten soon anyway.

**Comment by losipiuk** [comment]
As for broken `round` there were bugs concerning crossing short/long boundary. Both for round and roundN. I added tests and hopefully fixes.

**Comment by losipiuk** [comment]
Truncate is also broken :/. I am looking into that.

File: presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java (hunk 16 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java]: removed

**Comment by losipiuk** [comment]
@haozhun 
I applied all the comments and fixed/added better testing for `round`/`roundN`. 
I think it is ready for next review round.
Of course feel free to merge parts of this PR if you think they are ready.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 17]
```diff
@@ -682,6 +686,187 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
...
+        }
+        if (num.signum() < 0) {
+            return encodeUnscaledValue(roundLongLongDecimal(num.negate(), rescaleFactor).negate());
+        }
+        else {
+            return encodeUnscaledValue(roundLongLongDecimal(num, rescaleFactor));
+        }
+    }
+
+    @UsedByGeneratedCode
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: This method is not used by generated code directly.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 18]
```diff
@@ -682,6 +686,187 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
...
+
+    private static List<Object> decimalRoundNExtraParameters(SpecializeContext context)
+    {
+        return ImmutableList.of(context.getLiteral("num_precision"), context.getLiteral("num_scale"));
+    }
+
+    @UsedByGeneratedCode
+    public static long roundNShortShortDecimal(long num, long roundScale, long inputPrecision, long inputScale)
+    {
+        if (num == 0 || inputPrecision - inputScale + roundScale <= 0) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: I suppose `inputPrecision - inputScale + roundScale <= 0` should apply to short-long and long-long as well.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 19]
```diff
@@ -682,6 +686,187 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(S
...
+            return encodeUnscaledValue(0);
+        }
+        BigInteger rescaleFactor = bigIntegerTenToNth((int) (inputScale - roundScale));
+        if (unscaledValue.signum() < 0) {
+            return encodeUnscaledValue(roundNLongDecimal(unscaledValue.negate(), rescaleFactor).negate());
+        }
+        return encodeUnscaledValue(roundNLongDecimal(unscaledValue, rescaleFactor));
+    }
+
+    public static BigInteger roundNLongDecimal(BigInteger num, BigInteger rescaleFactor)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: private

**Comment by haozhun** [comment]
I find the similarities and differences between floor/ceil/round(N)/trunc(N) confusing. So I made up this table.

|        | can overflow | short can become  | long can become |
| ------ | :----------: | :---------------: | :-------------: |
| floor  | no           | short             | short, long     |
| ceil   | no           | short             | short, long     |
| round  | no           | short             | short, long     |
| roundN | yes          | short, long       | long            |
| trunc  | no           | short             | short, long     |
| truncN | no           | short             | long            |

I also went though this pull request to look at how things are handled. It looks like there are issues:

|        | `min(38,...)` for precision | runtime check for overflow |
| ------ | :-------------------------: | :------------------------: |
| floor  | no                          | no                         |
| ceil   | no                          | no                         |
| round  | yes (needs fix)             | no                         |
| roundN | yes                         | yes                        |
| trunc  | N/A                         | N/A                        |
| truncN | no                          | no                         |

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 17 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 18 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: It is actually checked in `private static Slice roundNLongDecimal` called by both.

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: I also spotted a bug in `roundNShortLongDecimal`
```
        if (roundScale >= inputScale) {
            return encodeUnscaledValue(ZERO);
        }
```

should be:
```
        if (roundScale >= inputScale) {
            return encodeUnscaledValue(num);
        }
```
But tests did not catch this.
I will add tests and fix.

**Comment by haozhun** [comment]
It looks like the `sign` function disappeared from this pull request.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 18 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: fixec

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 19 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: fixed

**Comment by losipiuk** [comment]
> It looks like the sign function disappeared from this pull request.

This is weird :/. I found it via `git reflog`. Hopefully I did not loose something else. 
It would nice to have PR history on github.

**Comment by haozhun** [comment]
I think this pull request looks good. I will be able to merge once you put up your new changes.

In my very long comment above, I mentioned that `min(38, ...)` shouldn't be necessary for declaration of `round`. Please fix that as well.

**Comment by losipiuk** [comment]
Yeah - I understood :). I removed min(38,) from type expression.

**Comment by losipiuk** [comment]
@haozhun 

I applied your comments + did some cleanup in commit messages.
I left the fixup commit unsquashed so it is easier for you to see what got changed.
There is a commit which reintroduces `sign` scalar :)

Let's wait for travis build.

Thanks for good review.

**Comment by haozhun** [comment]
`class Abs` and `class Sign` should have private constructors to forbid construction.

**Comment by haozhun** [comment]
Merged #5378.
